Add support for scalars and numpy in torch.testing

pmeier opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #56373 add `torch.testing` to docs
* #56365 Split `test_testing.py::TestAsserts` for multiple devices
* #55890 enable support for custom error messages in `torch.testing`
* #55788 add support for equal_nan in torch.testing.assert_close
* **#55786 Add support for scalars and numpy in torch.testing**
* #55385 Add support for checking tensor containers in `torch.testing`

Add support to compare scalars as well as `np.ndarray`'s with torch.testing. We are reusing the mathcing functionality that is already in place for tensors, by casting the inputs. The approach can easily extended if we want to support other input types as long as they can be cast to a tensor.

Differential Revision: [D27903814](https://our.internmc.facebook.com/intern/diff/D27903814)

pmeier committed:
Add support for scalars and numpy in torch.testing

Add support to compare scalars as well as `np.ndarray`'s with torch.testing. We are reusing the mathcing functionality that is already in place for tensors, by casting the inputs. The approach can easily extended if we want to support other input types as long as they can be cast to a tensor.

[ghstack-poisoned]

pmeier committed:
Update on "Add support for scalars and numpy in torch.testing"


Add support to compare scalars as well as `np.ndarray`'s with torch.testing. We are reusing the mathcing functionality that is already in place for tensors, by casting the inputs. The approach can easily extended if we want to support other input types as long as they can be cast to a tensor.

[ghstack-poisoned]

pmeier committed:
Update on "Add support for scalars and numpy in torch.testing"


Add support to compare scalars as well as `np.ndarray`'s with torch.testing. We are reusing the mathcing functionality that is already in place for tensors, by casting the inputs. The approach can easily extended if we want to support other input types as long as they can be cast to a tensor.

[ghstack-poisoned]

pmeier committed:
Update on "Add support for scalars and numpy in torch.testing"


Add support to compare scalars as well as `np.ndarray`'s with torch.testing. We are reusing the mathcing functionality that is already in place for tensors, by casting the inputs. The approach can easily extended if we want to support other input types as long as they can be cast to a tensor.

[ghstack-poisoned]

pmeier committed:
Update on "Add support for scalars and numpy in torch.testing"


Add support to compare scalars as well as `np.ndarray`'s with torch.testing. We are reusing the mathcing functionality that is already in place for tensors, by casting the inputs. The approach can easily extended if we want to support other input types as long as they can be cast to a tensor.

[ghstack-poisoned]

pmeier committed:
Update on "Add support for scalars and numpy in torch.testing"


Add support to compare scalars as well as `np.ndarray`'s with torch.testing. We are reusing the mathcing functionality that is already in place for tensors, by casting the inputs. The approach can easily extended if we want to support other input types as long as they can be cast to a tensor.

[ghstack-poisoned]

mattip commented:
Are there corner cases where casting an ndarray or a scalar to a Tensor is not the right thing to do? For instance, comparing a complex-valued scalar to an ndarray succeeds but casting them to Tensors and them comparing them fails?

mruberry commented on torch/testing/_asserts.py:
```diff
@@ -498,20 +468,65 @@ def _check_mapping(
+    """Maybe casts the input to a tensor.
+
+    :class:`~torch.Tensor`'s are returned without modification. If :mod:`numpy` is available, :class:`numpy.ndarray`'s
+    are cast with :func:`torch.from_numpy`. Everything else is casted with :func:`torch.tensor`.
```
"casted" -> "cast"

mruberry commented on torch/testing/_asserts.py:
```diff
@@ -649,7 +666,7 @@ def assert_close(
     """
     check_tensors = functools.partial(
-        _check_tensors_close,
+        _cast_inputs(_check_tensors_close),
```
I'm not in love with the idea that scalars, lists, arrays and tensors can be "close." 

Arrays and tensors make sense to me (although we should be sure that if a NumPy array is on the CPU and a tensor is on a CUDA device then check_device must be False), but if we want to cast other types to tensors maybe we need a different function or flag.

Note that in general this approach won't work for supporting scalar <-> tensor or list <-> tensor comparisons because it will also require check_device=False to work with CUDA tensors (as mentioned above). So a more permissive function that also doesn't check device would seem appropriate for this case.

mruberry replied:
(Not that I think this case is necessarily worth supporting. I generally wouldn't want engineers to use a more permissive cross-type version because it'd be harder to verify that the operation was returning the correct type of object.)

pmeier replied:
My idea behind all this was to support three basic types:

1. `torch.Tensor`,
2. `np.ndarray`, and
3. python scalars (`int`, `float`, and later `complex`)

They are simply cast into a tensor to use the same machinery for everything. Furthermore, I want to support sequences and mappings of any of the valid types, which will be checked elementwise.

> I'm not in love with the idea that scalars, lists, arrays and tensors can be "close."
>
> Arrays and tensors make sense to me [...], but if we want to cast other types to tensors maybe we need a different function or flag.

- **scalars**: Not sure why a scalar shouldn't be able to be close to a tensor. What is wrong with checking `torch.testing.assert_close(torch.tensor(1.0), 1.0)`?
- **lists**: I assume you mean lists of scalars, correct? This is basically a byproduct, because I didn't enforce only the three types above. So right now you could do `torch.testing.assert_close([1.0, 2.0], torch.tensor([1.0, 2.0]))`. If we don't want that, bailing out in such a scenario can be added.

> although we should be sure that if a NumPy array is on the CPU and a tensor is on a CUDA device then check_device must be False

I haven't actually thought about that. The same would be needed for `check_dtype` if we allow scalars. We could have them as `Optional[bool] = None` and set this depending on the situation.

Thoughts?

mruberry commented on torch/testing/_asserts.py:
```diff
@@ -379,43 +367,29 @@ def _check_tensors_close(
-            In case they mismatch should return an :class:`Exception` with an expressive error message.
+        actual (Any): Actual input.
+        expected (Any): Expected input.
+        check_data (Callable[[Any, Any], Optional[Exception]]): Callable used to check if a data pair matches.
```
"data pair" is not defined

mruberry commented on torch/testing/_asserts.py:
```diff
@@ -498,20 +468,65 @@ def _check_mapping(
+        b_t = _maybe_to_tensor(expected)
+        if a_t is None or b_t is None:
+            return UsageError(
+                f"Both inputs have to be tensors or something that can be cast to one, "
```
Not cast to a tensor but used to construct one from torch.tensor(), right?

pearu replied:
Or rather `torch.as_tensor` that avoids copies for certain types of inputs (tensors, ndarrays, other array-like objects implementing Array Interface, etc)?

pmeier commented:
Before we continue our implementation review cycle, we should probably define what kind of inputs we allow.

My idea is to allow all "array-and-scalar-like" inputs. This includes tensors as well as every other object from which a tensor can be created with `torch.tensor()` (see my comments about `check_device` and `check_dtype` at the end). This enables us to use our internal tensor comparison functionality for everything.

I would also allow sequences and mappings of the former types ("container of array-and-scalar-likes") for both inputs simultaneously. In this case the checks are performed elementwise.

I think this is a straight-forward UI in all but one case:
```
assert_close(torch.tensor([[1.0, 2.0], [3.0, 4.0]]), [[1.0, 2.0], [4.0, 4.0]]) # compare 2d tensors
assert_close([[1.0, 2.0], [3.0, 4.0]], [[1.0, 2.0], [4.0, 4.0]]) # compare sequence of 1d tensors
```

In the first case the error message would point to index `(1, 0)` whereas in the second case the error message would point to index `0` with the failure happening for index `1`. The information is basically the same, but it in the second case it is presented more convoluted. 

Still, IMO the special case of comparing two multi-level sequences of scalars does not outweigh the benefits of this more general approach.

---

All of this requires `check_device` and `check_dtype` to be flexible: 
- If we only deal with `torch.Tensor`'s', we can be strict and set `check_device = check_dtype = True` as it is the default now.
- If we additionally encounter an array-like, we should set `check_device = False`, but keep `check_dtype = True`, to avoid failures when comparing a CUDA tensor to the array.
- If we additionally encounter a scalar we should relax both checks `check_device = check_dtype = False`, since in addition to CUDA tensors the dtype of `torch.tensor()` is flexible and should not be checked if we internally construct the tensor from the inputs.

I would simply set the default value to `None` and in case it is not explicitly passed use the logic above to determine the default value.

pearu commented on torch/testing/_asserts.py:
```diff
@@ -30,6 +31,17 @@ class UsageError(Exception):  # type: ignore[no-redef]
 
+# 'numpy' is not a hard requirement of 'torch'. If it is available, we can import the underlying data directly with
+# 'torch.from_numpy()'. Otherwise we treat 'numpy.ndarray's the same as any other unknown object.
+try:
+    import numpy as np
+
+    _NUMPY_AVAILABLE = True
+except ImportError:
+    np = None  # type: ignore[assignment]
+    _NUMPY_AVAILABLE = False
```
`torch.as_tensor` is equivalent to `np.from_numpy`. I suggest replacing `from_numpy` usages with `as_tensor` to eliminate numpy references here as not needed.

mruberry commented:
> Before we continue our implementation review cycle, we should probably define what kind of inputs we allow.
> 
> My idea is to allow all "array-and-scalar-like" inputs. This includes tensors as well as every other object from which a tensor can be created with `torch.tensor()` (see my comments about `check_device` and `check_dtype` at the end). This enables us to use our internal tensor comparison functionality for everything.
> 
> I would also allow sequences and mappings of the former types ("container of array-and-scalar-likes") for both inputs simultaneously. In this case the checks are performed elementwise.
> 
> I think this is a straight-forward UI in all but one case:
> 
> ```
> assert_close(torch.tensor([[1.0, 2.0], [3.0, 4.0]]), [[1.0, 2.0], [4.0, 4.0]]) # compare 2d tensors
> assert_close([[1.0, 2.0], [3.0, 4.0]], [[1.0, 2.0], [4.0, 4.0]]) # compare sequence of 1d tensors
> ```
> 
> In the first case the error message would point to index `(1, 0)` whereas in the second case the error message would point to index `0` with the failure happening for index `1`. The information is basically the same, but it in the second case it is presented more convoluted.
> 
> Still, IMO the special case of comparing two multi-level sequences of scalars does not outweigh the benefits of this more general approach.
> 
> All of this requires `check_device` and `check_dtype` to be flexible:
> 
> * If we only deal with `torch.Tensor`'s', we can be strict and set `check_device = check_dtype = True` as it is the default now.
> * If we additionally encounter an array-like, we should set `check_device = False`, but keep `check_dtype = True`, to avoid failures when comparing a CUDA tensor to the array.
> * If we additionally encounter a scalar we should relax both checks `check_device = check_dtype = False`, since in addition to CUDA tensors the dtype of `torch.tensor()` is flexible and should not be checked if we internally construct the tensor from the inputs.
> 
> I would simply set the default value to `None` and in case it is not explicitly passed use the logic above to determine the default value.

From offline discussion:

- let's require types match, except for containers where it's OK for sequences to compare with other sequences and mappables to compared with other mappables (for example, a list and a tuple with the same elements would still be equal under this definition, as would a dict and an ordered dict with the same keys and values)
- converting to tensor in advance sounds good, but we should be sure that error messages for scalars and 0D tensors (AKA "scalar tensors") reflect the user's expectations and don't use language like "tensors failed to compare equal at index 0", because the user may not have compared tensors and scalars and 0D tensors don't support indexing

Note this means that [1, 2] != torch.tensor((1, 2)) and that 2 != complex(2, 0).

pmeier committed:
Update on "Add support for scalars and numpy in torch.testing"


Add support to compare scalars as well as `np.ndarray`'s with torch.testing. We are reusing the mathcing functionality that is already in place for tensors, by casting the inputs. The approach can easily extended if we want to support other input types as long as they can be cast to a tensor.

[ghstack-poisoned]

pmeier commented:
@mruberry Should be all done except for 

> error messages for scalars and 0D tensors (AKA "scalar tensors") reflect the user's expectations and don't use language like "tensors failed to compare equal at index 0", because the user may not have compared tensors and scalars and 0D tensors don't support indexing

I agree that we should do that, but lets wait with that until we figured out what to do with #55890.

mruberry commented:
> @mruberry Should be all done except for
> 
> > error messages for scalars and 0D tensors (AKA "scalar tensors") reflect the user's expectations and don't use language like "tensors failed to compare equal at index 0", because the user may not have compared tensors and scalars and 0D tensors don't support indexing
> 
> I agree that we should do that, but lets wait with that until we figured out what to do with #55890.

Sounds good; I'll take a look later today

pmeier committed:
Update on "Add support for scalars and numpy in torch.testing"


Add support to compare scalars as well as `np.ndarray`'s with torch.testing. We are reusing the mathcing functionality that is already in place for tensors, by casting the inputs. The approach can easily extended if we want to support other input types as long as they can be cast to a tensor.

[ghstack-poisoned]

mruberry commented:
This is a tricky stack to parse.

So let me just ask a few questions and then maybe we can revisit this code after this stack. What will happen with the following tests:

```
np.array((1, 2)) == torch.tensor((1, 2))
np.array((1, 2)) == np.array((1, 2))
complex(2, 0) == 2
2. == 2
[1, 2] == torch.tensor((1, 2))
2. == torch.tensor((2.,))
2. == np.array((2.,))
```

pmeier commented:
@mruberry Assuming `==` means `torch.testing.assert_(equal|close)`:

- `np.array((1, 2)) == torch.tensor((1, 2))`: Fail, because the inputs have a different type and we required type equality.
- `np.array((1, 2)) == np.array((1, 2))`: Pass.
- `complex(2, 0) == 2`: Fail, because we don't support complex tensors right now and `torch.as_tensor(complex(2, 0)).dtype == torch.complex64`
- `2. == 2`: Fail, because the inputs have a different type and we required type equality.
- `[1, 2] == torch.tensor((1, 2))`: Fail, because the inputs have a different type and we required type equality.
- `2. == torch.tensor((2.,))`: Fail, because the inputs have a different type and we required type equality.
- `2. == np.array((2.,))`: Fail, because the inputs have a different type and we required type equality.

mruberry reviewed (approved):
Cool!

pmeier committed:
Update on "Add support for scalars and numpy in torch.testing"


Add support to compare scalars as well as `np.ndarray`'s with torch.testing. We are reusing the mathcing functionality that is already in place for tensors, by casting the inputs. The approach can easily extended if we want to support other input types as long as they can be cast to a tensor.

[ghstack-poisoned]

pmeier committed:
Update on "Add support for scalars and numpy in torch.testing"


Add support to compare scalars as well as `np.ndarray`'s with torch.testing. We are reusing the mathcing functionality that is already in place for tensors, by casting the inputs. The approach can easily extended if we want to support other input types as long as they can be cast to a tensor.

Differential Revision: [D27903814](https://our.internmc.facebook.com/intern/diff/D27903814)

[ghstack-poisoned]

pmeier committed:
Update on "Add support for scalars and numpy in torch.testing"


Add support to compare scalars as well as `np.ndarray`'s with torch.testing. We are reusing the mathcing functionality that is already in place for tensors, by casting the inputs. The approach can easily extended if we want to support other input types as long as they can be cast to a tensor.

Differential Revision: [D27903814](https://our.internmc.facebook.com/intern/diff/D27903814)

[ghstack-poisoned]

pmeier committed:
Update on "Add support for scalars and numpy in torch.testing"


Add support to compare scalars as well as `np.ndarray`'s with torch.testing. We are reusing the mathcing functionality that is already in place for tensors, by casting the inputs. The approach can easily extended if we want to support other input types as long as they can be cast to a tensor.

Differential Revision: [D27903814](https://our.internmc.facebook.com/intern/diff/D27903814)

[ghstack-poisoned]
